### PR TITLE
Bind transport error

### PR DESCRIPTION
**Pull request by zippy**
## PR summary

Handle unbinding events and update transport error to handle this:

- renames Disconnected event in Lib3hToClient to Unbound (Still mapping this to Disconnected in Lib3hServerProtocol for now, but nobody's using that so it will go away).
- Adds ErrorKind to TransportError (like our other error types)
  - adds Unbound error kind
  - adds Ignore error kind
- propagates ErrorOccured transport errors up the stack and converts to Unbound when appropriate

Followup:

- [ ] figure out which tungstenite error is returned with a acceptor becomes unbound.

## Review checklist 
- [ ] The story has unit or integration tests
- [ ] No new bugs, and any tech-debt is identified and justified
- [ ] There is enough API documentation (how to use)
- [ ] There is enough code documentation (how the code works)

### CODE REVIEW

File: crates/lib3h/src/engine/network_layer.rs [hunk 1]
```diff
@@ -119,15 +120,23 @@ impl<'engine> GhostEngine<'engine> {
...
-                    let data = DisconnectedData {
-                        network_id: "FIXME".to_string(), // TODO #172
-                    };
+                if error.kind() == &transport::error::ErrorKind::Unbind {
+                    let data = UnboundData { uri: uri.clone() };
                     self.lib3h_endpoint
-                        .publish(Span::fixme(), Lib3hToClient::Disconnected(data))?;
+                        .publish(Span::fixme(), Lib3hToClient::Unbound(data))?;
+                } else {
+                    panic!("unhandled error {}", error);
```
Comment (struktured) [inline crates/lib3h/src/engine/network_layer.rs]: Is this intended behavior or a follow up is necessary?

File: crates/lib3h/src/transport/error.rs [hunk 2]
```diff
@@ -10,98 +10,144 @@ pub type TransportResult<T> = Result<T, TransportError>;
 
 /// represents an error generated by a connection instance
 #[derive(Debug, PartialEq, Clone)]
-pub struct TransportError(pub String);
+pub struct TransportError(Box<ErrorKind>);
```
Comment (struktured) [inline crates/lib3h/src/transport/error.rs]: Surprsied to see a boxed type here- not a huge deal, but what precipitated it?

File: crates/lib3h_protocol/src/protocol.rs [hunk 3]
```diff
@@ -292,8 +294,10 @@ impl From<Lib3hToClient> for Lib3hServerProtocol {
             Lib3hToClient::Connected(connected_data) => {
                 Lib3hServerProtocol::Connected(connected_data)
             }
-            Lib3hToClient::Disconnected(disconnected_data) => {
-                Lib3hServerProtocol::Disconnected(disconnected_data)
+            Lib3hToClient::Unbound(_unbound_data) => {
+                Lib3hServerProtocol::Disconnected(DisconnectedData {
+                    network_id: "".into(),
```
Comment (struktured) [inline crates/lib3h_protocol/src/protocol.rs]: Don't quite understand this conversion.

File: crates/lib3h/src/transport/error.rs (hunk 2 above)
Comment (zippy) [reply crates/lib3h/src/transport/error.rs]: I was just copying how this was done in our other error types.  Do you think we should remove it everywhere?

File: crates/lib3h_protocol/src/protocol.rs (hunk 3 above)
Comment (zippy) [reply crates/lib3h_protocol/src/protocol.rs]: It's a good question.  Currently Disconnected is not being used by anything or in core, but it's the place this message was originally connected to in n3h (I think) so I reconnected it.  It should go away.

File: crates/lib3h/src/transport/error.rs (hunk 2 above)
Comment (struktured) [reply crates/lib3h/src/transport/error.rs]: For this PR just leave it alone.

File: crates/lib3h/bin/send-demo.rs [hunk 4]
```diff
@@ -35,19 +36,31 @@ struct EngineContainer<
...
 impl<'lt> EngineContainer<GhostEngine<'lt>> {
-    pub fn new() -> Self {
+    pub fn new(ws: bool) -> Self {
         let crypto = Box::new(SodiumCryptoSystem::new());
 
-        let config = EngineConfig {
+        let (transport_config, bind_url) = if ws {
+            (
+                TransportConfig::Websocket(TlsConfig::Unencrypted),
+                Url::parse("ws://127.0.0.1:63518").unwrap(),
```
Comment (neonphog) [inline crates/lib3h/bin/send-demo.rs]: out of curiosity, does `ws://127.0.0.1:0` work?

File: crates/lib3h/src/transport/error.rs [hunk 5]
```diff
@@ -10,98 +10,144 @@ pub type TransportResult<T> = Result<T, TransportError>;
...
+        &self.0
+    }
+
+    /// Unwrap this error into its underlying type.
+    pub fn into_kind(self) -> ErrorKind {
+        *self.0
+    }
+
+    /// is this an Ignore kind error
+    pub fn is_ignorable(&self) -> bool {
```
Comment (neonphog) [inline crates/lib3h/src/transport/error.rs]: : ) :+1:

File: crates/lib3h/src/engine/network_layer.rs (hunk 1 above)
Comment (zippy) [reply crates/lib3h/src/engine/network_layer.rs]: removed panic and added issue #391

**Pull request merged**